### PR TITLE
Try first cut at issue 368

### PR DESCRIPTION
**Pull request by ckormanyos**
This PR attempts to address #368 with minimally non-intrusive changes of rounding in selected (but not all) places.

### CODE REVIEW

**Comment by ckormanyos** [comment]
Hi @jzmaddock what's your opinion on this cut?

I introduced rounding-up in selected areas only including `floor`/`ceil`/cast-to-built-in-integral.

Advantages include:
- Handle the matter of #368 and possibly a subclass of related issues.
- Introduce some rounding into `cpp_dec_float` to be expanded later.
- Gain experience.

Disadvantage is that rounding has not yet been successfully applied to arithmetic operations and there are some open questions regarding the best way to do this. This could/should be completed within the framework of also refining equality on the sub-limb level.

At the moment, however, there are no tangible disadvantages other than items that were already missing anyway. And CI is passing (except Circle CI has been spinning around for a day or so).

So what do you think of this PR, its scope of change and if we should actually merge?

 is

File: test/git_issue_368.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      {
+        const std::uint32_t lo_hi = lo_index * hi_index;
+
+        const decimal_type a = decimal_type { lo_hi } / decimal_type { lo_index };
+        const decimal_type b = ceil(a);
+
+        const std::string str_a = boost::lexical_cast<std::string>(a);
+        const std::string str_b = boost::lexical_cast<std::string>(b);
+
+        result_is_ok &= (str_a == str_b);
```
Comment (jzmaddock) [inline test/git_issue_368.cpp:31]: Questions: why are we comparing strings rather than values, and should we also test inputs which are not integer?  Negative values as well, plus floor too I suspect ;)

File: include/boost/multiprecision/cpp_dec_float.hpp [hunk 2]
```diff
@@ -3091,37 +3236,73 @@ template <unsigned Digits10, class ExponentType, class Allocator>
...
+   else
+   {
+      result.eval_round_self();
 
-   if (x.isneg())
-      result -= cpp_dec_float<Digits10, ExponentType, Allocator>::one();
-   result = result.extract_integer_part();
+      if(result.isint())
+      {
+         ;
```
Comment (jzmaddock) [inline include/boost/multiprecision/cpp_dec_float.hpp:3251]: ?

File: include/boost/multiprecision/cpp_dec_float.hpp [hunk 3]
```diff
@@ -522,6 +522,141 @@ class cpp_dec_float
...
+      if(needs_rounding)
+      {
+         using local_size_type = typename array_type::size_type;
+         using local_limb_type = typename array_type::value_type;
+
+         std::int32_t digits_limb_0 = 0;
+
+         local_limb_type tmp_limb_0 = data[0U];
+
+         // Manually count the number of base-10 digits on the zero'th limb.
```
Comment (jzmaddock) [inline include/boost/multiprecision/cpp_dec_float.hpp:538]: This is the bit that caused me to assume that correct round was next to impossible to do efficiently in this type, but maybe it's not so bad?  I can't help but wonder if there are more efficient implementations - maybe binary search comparisons?  Or can we even do table lookup based on the most-significant-bit?  Just thinking out loud here....

File: test/git_issue_368.cpp (hunk 1 above)
Comment (ckormanyos) [reply test/git_issue_368.cpp]: > why are we comparing strings rather than values, 

In this PR, there is rounding on `floor`, `ceil` and cast-to-int only. I removed the use of strings. But the comparison with, say, the result of `floor` or `ceil` is only true if the value being compared is casted to int prior to comparison. This is now done without strings in the candidate code of the PR's test file.

> we also test inputs which are not integer

It seems like there are already-existing rounding tests which do this.

> Negative values as well, plus floor too I suspect

Understood. Done in resent push(es).

File: include/boost/multiprecision/cpp_dec_float.hpp (hunk 2 above)
Comment (ckormanyos) [reply include/boost/multiprecision/cpp_dec_float.hpp]: > `if(result.isint())`
> ?

Indeed. These sequences have been simlified in recent push(es).

File: include/boost/multiprecision/cpp_dec_float.hpp (hunk 3 above)
Comment (ckormanyos) [reply include/boost/multiprecision/cpp_dec_float.hpp]: > the bit that caused me to assume that correct round was next to impossible to do efficiently in this type

Well, it sure is proving to be unruly. But there are some new efficiency provisions.
- There was an existing binary search which had calculated the number of base-10 digits in the leading limb. This is used instead of a manual counting loop now.
- For the power-of-10, a special non-`constexpr` table lookup has also been created for quick calculations of the power(s) of ten in similar matters.

**Comment by ckormanyos** [comment]
Hi @jzmaddock I have attempted to address the review questions one-by-one.

Personally, I am skeptical to introduce _partial_ rounding on `floor`, `ceil`, and cast-to-int only.

The questions I ask myself are:

1. Is `cpp_dec_float` worse off with partial rounding?
2. Is `cpp_dec_float` neutral or about the same?
3. Or does this particl rounding improve the overall class without high risk of breaking anything else?

I tend to think we are around 2 and 3. I would, however, treat this skeptically. We deal with #368. But it seems like a partial treatment of rounding, and partial could be interpreted as wrong or less right than uniform truncation.

I ran some tests with rounding on arithmetic operatory, but found that this leads to quite a few changes in tolerances and behaviors that are already possibly running. I am not sure if rounding over all (also on arithmetic operations) would break existing code or not? I also came to believe that rounding only makes sense if we combine it with a more exact reatment of equality (equality on the sub-limb, digit level). This has also been avoided over the years.

I have no problem keeping these changes in the PR or letting it sit for a while, or even dropping this effort. I am, however, somewhat motivated by the urge to deal with #368.

Thoughts?
